frame/staking/src/lib.rs

```diff
@@ -761,6 +766,102 @@ impl<AccountId, Balance: HasCompact + Zero> UnappliedSlash<AccountId, Balance> {
 	}
 }
 
+/// Something that defines the maximum number of nominations per nominator based on a curve.
+pub trait NominationsQuota<Balance> {
+	/// Maximum number of nominations. The method `curve` implements the nomination quota curve and
+	/// should not be used directly. However, `get_quota` returns the bounded maximum number of
+	// nominations based on `fn curve` and the nominator's balance.
+	type MaxNominations: Get<u32>;
+
+	/// Returns the voter's nomination quota within reasonable bounds [`min`, `max`], where `min`
+	/// is 1 and `max` is `Self::MaxNominations`.
+	fn get_quota(balance: Balance) -> u32 {
+		Self::curve(balance).clamp(1, Self::MaxNominations::get())
+	}
+
+	// Returns the voter's nomination quota based on its balance and a curve.
+	fn curve(balance: Balance) -> u32;
+}
+
+/// A nomination quota that allows up to MAX nominations for all validators.
+pub struct FixedNominationsQuota<const MAX: u32>;
+impl<Balance, const MAX: u32> NominationsQuota<Balance> for FixedNominationsQuota<MAX> {
+	type MaxNominations = Self;
+
+	fn curve(_: Balance) -> u32 {
+		MAX
+	}
+}
+
+impl<const MAX: u32> Get<u32> for FixedNominationsQuota<MAX> {
+	fn get() -> u32 {
+		MAX
+	}
+}
+
+/// A static tracker for the election data snapshot.
+///
+/// Computes the (SCALE) encoded byte length of a snapshot based on static rules, without any actual
+/// encoding.
+///
+/// ## Details
+///
+/// The snapshot has the form of `Vec<Voter>` where `Voter = (Account, u64, Vec<Account>)`. For
+/// each voter added to the snapshot, [`try_register_voter`] should be called, with the number
+/// of votes (length of the internal `Vec`).
+///
+/// Whilst doing this, [`size`] will track the entire size of the `Vec<Voter>`, except for the
+/// length prefix of the outer `Vec`. To get the final size at any point, use
+/// [`final_byte_size_of`].
+pub(crate) struct ElectionSizeTracker<AccountId> {
+	pub size: usize,
+	_marker: sp_std::marker::PhantomData<AccountId>,
+}
+
+impl<AccountId> ElectionSizeTracker<AccountId> {
```

Comment:
would like to see independent tests for this, I recall original PR had it.

Reply:
There are unit tests for the `ElectionSizeTracker` here https://github.com/paritytech/substrate/pull/12970/files#diff-64c049b4e94a55bdeaf757c725cf7d1df623b754557af4ced157024a943c4703R5908 and the boundaries are also implicitly tested across other unit tests. I'll see if the test coverage can be improved though.

Reply:
Check the previous tests, add more coverage and check the size_hint option.